# Vorlage Radius

In [1]:
import random
import networkx 
import itertools
from collections import defaultdict

In [17]:

import random
import networkx
import itertools
from collections import defaultdict

def random_answers(n_participants, n_questions):
    participants=[[random.randrange(2) for j in range(n_questions)] for i in range(n_participants)]
    return participants
    
def create_graph(answers, n_participants, n_questions, threshold):
    G = networkx.Graph()
    G.add_nodes_from(list(range(n_participants)))
    for i, j in itertools.combinations(range(n_participants), 2):
        n_disagreements = sum(answers[i][k] ^ answers[j][k] for k in range(n_questions))
        if n_disagreements >= threshold:
            G.add_edge(i, j)
    return G

def test_weighted_match(n_participants, n_questions, threshold):
    A = random_answers(n_participants, n_questions)
    G = create_graph(A, n_participants, n_questions, threshold)
    matching = networkx.max_weight_matching(G)
    
    return len(matching)

def get_freqs(n_participants, n_questions, threshold, n_runs):
    freq = defaultdict(lambda: 0.0)
    for i in range(n_runs):
        res = test_weighted_match(n_participants, n_questions, threshold)
        freq[res] += 1/n_runs
    return freq

if __name__ == "__main__":
    n_runs = 1000
    n_questions = 5
    thresholds = [3,4]
    group_sizes = [10, 20, 30, 40, 50, 70, 90]

    for threshold in thresholds:
        for n_participants in group_sizes:
            print(f"{threshold} out of {n_questions}, group size {n_participants}\n")
            print("matches | frequency\n")
            for k, v in sorted(list(get_freqs(n_participants, n_questions, threshold, n_runs).items())):
                print(f"     {k:2} | {v:6.1%} |{'='*int(v*50):50s}|")
            print()

3 out of 5, group size 10

matches | frequency

      2 |   0.2% |                                                  |
      3 |   2.1% |=                                                 |
      4 |  12.7% |======                                            |
      5 |  85.0% |==========================================        |

3 out of 5, group size 20

matches | frequency

      9 |   0.9% |                                                  |
     10 |  99.1% |================================================= |

3 out of 5, group size 30

matches | frequency

     15 | 100.0% |==================================================|

3 out of 5, group size 40

matches | frequency

     20 | 100.0% |==================================================|

3 out of 5, group size 50

matches | frequency

     25 | 100.0% |==================================================|

3 out of 5, group size 70

matches | frequency

     35 | 100.0% |==================================================|

3 out 

3 out of 5, group size 10

matches | frequency

      2 |   0.2% |                                                  |
      3 |   2.1% |=                                                 |
      4 |  12.7% |======                                            |
      5 |  85.0% |==========================================        |

3 out of 5, group size 20

matches | frequency

      9 |   0.9% |                                                  |
     10 |  99.1% |================================================= |

3 out of 5, group size 30

matches | frequency

     15 | 100.0% |==================================================|

3 out of 5, group size 40

matches | frequency

     20 | 100.0% |==================================================|

3 out of 5, group size 50

matches | frequency

     25 | 100.0% |==================================================|

3 out of 5, group size 70

matches | frequency

     35 | 100.0% |==================================================|

3 out of 5, group size 90

matches | frequency

     45 | 100.0% |==================================================|

4 out of 5, group size 10

matches | frequency

      0 |   0.1% |                                                  |
      1 |   1.4% |                                                  |
      2 |  12.2% |======                                            |
      3 |  36.8% |==================                                |
      4 |  39.2% |===================                               |
      5 |  10.3% |=====                                             |

4 out of 5, group size 20

matches | frequency

      4 |   0.2% |                                                  |
      5 |   0.7% |                                                  |
      6 |   2.6% |=                                                 |
      7 |  10.6% |=====                                             |
      8 |  22.7% |===========                                       |
      9 |  38.1% |===================                               |
     10 |  25.1% |============                                      |

4 out of 5, group size 30

matches | frequency

      8 |   0.1% |                                                  |
      9 |   0.1% |                                                  |
     10 |   1.1% |                                                  |
     11 |   1.2% |                                                  |
     12 |   5.5% |==                                                |
     13 |  14.4% |=======                                           |
     14 |  28.5% |==============                                    |
     15 |  49.1% |========================                          |

4 out of 5, group size 40

matches | frequency

     16 |   1.2% |                                                  |
     17 |   2.9% |=                                                 |
     18 |   7.2% |===                                               |
     19 |  19.5% |=========                                         |
     20 |  69.2% |==================================                |

4 out of 5, group size 50

matches | frequency

     18 |   0.1% |                                                  |
     20 |   0.4% |                                                  |
     21 |   0.5% |                                                  |
     22 |   1.2% |                                                  |
     23 |   3.1% |=                                                 |
     24 |  11.7% |=====                                             |
     25 |  83.0% |=========================================         |

4 out of 5, group size 70

matches | frequency

     31 |   0.2% |                                                  |
     32 |   0.4% |                                                  |
     33 |   1.0% |                                                  |
     34 |   1.4% |                                                  |
     35 |  97.0% |================================================  |

4 out of 5, group size 90

matches | frequency

     42 |   0.2% |                                                  |
     43 |   0.3% |                                                  |
     44 |   0.6% |                                                  |
     45 |  98.9% |================================================= |

